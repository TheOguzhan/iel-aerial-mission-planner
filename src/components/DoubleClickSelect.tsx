import React from "react";

interface IDoubleClickSelectProps {
  options: String[];
  val: String;
  disabled?: boolean;
  onChange: Function;
  large?: boolean;
}

export const DoubleClickSelect: React.FunctionComponent<IDoubleClickSelectProps> =
  (props) => {
    const [select, setSelect] = React.useState<String>(props.val);
    const [edit, setEdit] = React.useState<boolean>(false);
    return (
      <>
        <div className="mt-1 mx-auto relative">
          <span
            className="flex items-center  select-none cursor-pointer dark:text-white text-black"
            onClick={(e) => setEdit(!edit)}
            data-testid="val-displayer"
          >
            {select}
          </span>
        </div>
        <ul
          className={`absolute z-10 mt-1 ${
            props.large ? "w-56" : "w-28"
          }  mx-auto bg-white shadow-lg max-h-56 rounded-md py-1 text-base ring-1 ring-black ring-opacity-5 overflow-auto focus:outline-none sm:text-sm`}
          tabIndex={-1}
          role="listbox"
          aria-labelledby="listbox-label"
          aria-activedescendant="listbox-option-3"
          style={{ display: edit && !props.disabled ? "block" : "none" }}
          onBlur={(e) => {
            setEdit(false);
          }}
          data-testid="main-select"
        >
          {props.options.map((element, index) => {
            return (
              <li
                className="text-gray-900  cursor-pointer select-none relative py-2 pl-2 pr-2 hover:bg-gray-200"
                id={`list-box-option-${index}`}
                role="option"
                aria-selected={select === element}
                onClick={() => {
                  setSelect(element);
                  props.onChange(element);
                  setEdit(false);
                }}
                key={index}
                data-testid="option"
              >
                <div className="flex items-items">
                  <span className="text-black text-font-normal ml-3 block truncate text-center">
                    {element}
                  </span>
                </div>
              </li>
            );
          })}
        </ul>
      </>
    );
  };

DoubleClickSelect.defaultProps = {
  disabled: false,
  large: false,
};
