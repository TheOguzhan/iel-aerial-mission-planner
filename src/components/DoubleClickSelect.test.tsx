import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import { DoubleClickSelect } from "./DoubleClickSelect";
import renderer from "react-test-renderer";

afterEach(cleanup);

describe("Double click select", () => {
  const options: Array<String> = ["option 1", "option 2", "option-3"];
  const val: String = "option 1";
  const onChange = () => {};

  render(<DoubleClickSelect options={options} val={val} onChange={onChange} />);

  const displayer = screen.getByTestId("val-displayer");
  const main = screen.getByTestId("main-select");
  const optionsElements = screen.getAllByTestId("option");

  test("should show displayer", () => {
    expect(displayer).toBeVisible();
    expect(main).not.toBeVisible();
    expect(optionsElements[0]).not.toBeVisible();
  });
  test("should match snapshots", () => {
    const tree = renderer.create(
      <DoubleClickSelect options={options} val={val} onChange={onChange} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
