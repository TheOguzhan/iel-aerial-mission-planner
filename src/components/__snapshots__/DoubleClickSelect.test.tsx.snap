// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Double click select should match snapshots 1`] = `
Array [
  <div
    className="mt-1 mx-auto relative"
  >
    <span
      className="flex items-center  select-none cursor-pointer dark:text-white text-black"
      data-testid="val-displayer"
      onClick={[Function]}
    >
      option 1
    </span>
  </div>,
  <ul
    aria-activedescendant="listbox-option-3"
    aria-labelledby="listbox-label"
    className="absolute z-10 mt-1 w-28  mx-auto bg-white shadow-lg max-h-56 rounded-md py-1 text-base ring-1 ring-black ring-opacity-5 overflow-auto focus:outline-none sm:text-sm"
    data-testid="main-select"
    onBlur={[Function]}
    role="listbox"
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
  >
    <li
      aria-selected={true}
      className="text-gray-900  cursor-pointer select-none relative py-2 pl-2 pr-2 hover:bg-gray-200"
      data-testid="option"
      id="list-box-option-0"
      onClick={[Function]}
      role="option"
    >
      <div
        className="flex items-items"
      >
        <span
          className="text-black text-font-normal ml-3 block truncate text-center"
        >
          option 1
        </span>
      </div>
    </li>
    <li
      aria-selected={false}
      className="text-gray-900  cursor-pointer select-none relative py-2 pl-2 pr-2 hover:bg-gray-200"
      data-testid="option"
      id="list-box-option-1"
      onClick={[Function]}
      role="option"
    >
      <div
        className="flex items-items"
      >
        <span
          className="text-black text-font-normal ml-3 block truncate text-center"
        >
          option 2
        </span>
      </div>
    </li>
    <li
      aria-selected={false}
      className="text-gray-900  cursor-pointer select-none relative py-2 pl-2 pr-2 hover:bg-gray-200"
      data-testid="option"
      id="list-box-option-2"
      onClick={[Function]}
      role="option"
    >
      <div
        className="flex items-items"
      >
        <span
          className="text-black text-font-normal ml-3 block truncate text-center"
        >
          option-3
        </span>
      </div>
    </li>
  </ul>,
]
`;
